util,changefeedccl: streaminger csv writes

HonoreDB opened the pull request:
The changefeed csv encoder was reading an entire row into memory before encoding it into CSV. There's no particular need to do so, and CSV is now being used at scale, so this PR modifies the CSV writer API slightly to allow for writing fields directly from an iterator.

Release justification: Performance tweak to new functionality.

Release note: None

HonoreDB committed:
util,changefeedccl: streaminger csv writes

The changefeed csv encoder was reading an entire row into memory
before encoding it into CSV. There's no particular need to do so,
and CSV is now being used at scale, so this PR modifies the CSV
writer API slightly to allow for writing fields directly from
an iterator.

Release justification: Performance tweak to new functionality.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87472)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
Is there a profile or benchmark that shows this mattering? I'd be particularly reticent to backport this change without evidence. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87472#-:-NBN7hyu0QUWBtIeSPP-:b-b0pyun)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava and @miretskiy)



<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87472)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

___
*[`pkg/ccl/changefeedccl/encoder_csv.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87472#-NBNCB1V3auY-HTBRfSh:-NBNCB1V3auY-HTBRfSi:b-f4k351) ([raw file](https://github.com/cockroachdb/cockroach/blob/6158b5c48f3f2e654c0c6523bf9b34fa810fd5b5/pkg/ccl/changefeedccl/encoder_csv.go#L30)):*
> ```Go
> var _ Encoder = &csvEncoder{}
> 
> func newCSVEncoder(opts changefeedbase.EncodingOptions) *csvEncoder {
> ```

We might as well remove the opts parameter since it's unused.


<!-- Sent from Reviewable.io -->

miretskiy commented:
Agree with @ajwerner ; This csv encoder is used in other parts (e.g. import), so benchmark would be good.
I actually doubt that this has that much of a difference; if we are worried about allocating array, then let's memoize row array; I don't think it would be necessary to change csv encoder.

HonoreDB committed:
changefeedccl, util: avoid allocs during csv encoding

BenchmarkCSVEncodeWideRow-16             	    2415	    484568 ns/op	   10131 B/op	    2304 allocs/op
BenchmarkCSVEncodeWideRowASCII-16        	    8277	    145277 ns/op	     262 B/op	       0 allocs/op
BenchmarkCSVEncodeWideColumnsASCII-16    	   15450	     77659 ns/op	     140 B/op	       0 allocs/op
BenchmarkCSVEncodeWideColumns-16         	    3168	    371453 ns/op	    9924 B/op	    2304 allocs/op

master:

BenchmarkCSVEncodeWideRow-16             	    2493	    762892 ns/op	   32710 B/op	    3328 allocs/op
BenchmarkCSVEncodeWideRowASCII-16        	    5935	    176015 ns/op	    8694 B/op	    1024 allocs/op
BenchmarkCSVEncodeWideColumnsASCII-16    	   14484	     81154 ns/op	    5528 B/op	       3 allocs/op
BenchmarkCSVEncodeWideColumns-16         	    2689	    442493 ns/op	   28486 B/op	    2307 allocs/op

This change reduces the memory usage of csv.Writer#WriteField by taking a bytes.Buffer instead
of a string, and never converting it into a string. To do this, we refactor the check for
whether the string needs to be enclosed in quotes to occur at the same time as the string
is being escaped, which works pretty naturally because there's a lot of overlap in the logic
required for quoting and escaping. The only behavior that this methodology can't easily support
is that if a custom escape character is used that is not otherwise a special character, we previously
only escaped it inside quoted strings, which here would require making another pass through after we
know whether or not we're quoting. However, it looks like custom escape characters aren't used, and
it's not clear that's the correct behavior anyway.

Most of the remaining memory usage comes from multibyte character processing. There's probably a way
to fix that if needed.

Release justification: Performance improvement to new functionality. CSV is being used for terabyte-level
exports and needs to be fast.

Release note: None

HonoreDB commented:
Added benchmarks (and improved the writer some more):
```
BenchmarkCSVEncodeWideRow-16             	    2362	    496653 ns/op	   10151 B/op	    2304 allocs/op
BenchmarkCSVEncodeWideRowASCII-16        	    7641	    143328 ns/op	     284 B/op	       0 allocs/op
BenchmarkCSVEncodeWideColumnsASCII-16    	   15202	     79449 ns/op	     142 B/op	       0 allocs/op
BenchmarkCSVEncodeWideColumns-16         	    3148	    376196 ns/op	    9928 B/op	    2304 allocs/op

master:

BenchmarkCSVEncodeWideRow-16             	    2493	    762892 ns/op	   32710 B/op	    3328 allocs/op
BenchmarkCSVEncodeWideRowASCII-16        	    5935	    176015 ns/op	    8694 B/op	    1024 allocs/op
BenchmarkCSVEncodeWideColumnsASCII-16    	   14484	     81154 ns/op	    5528 B/op	       3 allocs/op
BenchmarkCSVEncodeWideColumns-16         	    2689	    442493 ns/op	   28486 B/op	    2307 allocs/op
```

The idea here is to avoid allocating an entire copy of the row just to pass it from changefeedccl.csvEncoder to csv.Writer. Instead, we can keep the data in a single bytes.Buffer, one field at a time. This saves time (15-30% faster) and memory in each benchmark, but the key stat is that memory use goes from O(bytes) to O(number of multi-byte characters) in the input, with no new allocs when the row is only ASCII (really, only codepoints < 256 but I used < 128 in the benchmarks).

Y'all probably have a better intuition than me about how much this would help in practice with the ongoing effort to get a cdc export of a multi-terabyte table to happen in a reasonable amount of time.

I haven't taken advantage of the method I introduce here outside changefeeds in this PR (nor looked into how easy it would be to change the flow there). I think other callers will see a net performance improvement from only traversing each string once instead of thrice, but I haven't benchmarked that.

ajwerner reviewed (approved):
LGTM

HonoreDB commented:
bors r+

shermanCRL commented:
Looks like we intended/attempted a v22.1 backport. Worth pursuing @HonoreDB? Fine if yes, fine if no, just looking for clarity.
